internal/auth/internal/mappers/trafficpermissionsmapper/traffic_permissions_mapper.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package trafficpermissionsmapper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/auth/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/internal/resource/mappers/bimapper"
+	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type TrafficPermissionsMapper struct {
+	lock                     sync.Mutex
+	mapper                   *bimapper.Mapper
+	workloadIdentityToCTPMap map[string]*pbresource.ID
+}
+
+func New() *TrafficPermissionsMapper {
+	return &TrafficPermissionsMapper{
+		lock:                     sync.Mutex{},
+		mapper:                   bimapper.New(types.ComputedTrafficPermissionsType, types.TrafficPermissionsType),
+		workloadIdentityToCTPMap: make(map[string]*pbresource.ID, 0),
+	}
+}
+
+//  TrafficPermissionsMapper functions
+
+func (tm *TrafficPermissionsMapper) MapTrafficPermissions(_ context.Context, _ controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	mappedWIs := make([]controller.Request, 0)
+	// get already associated WorkloadIdentities/CTPs for reconcile requests:
+	associatedWIs := tm.mapper.ItemIDsForLink(res.Id)
+	for _, w := range associatedWIs {
+		mappedWIs = append(mappedWIs, controller.Request{ID: w})
+	}
+	var tp pbauth.TrafficPermissions
+	err := res.Data.UnmarshalTo(&tp)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&tp, err)
+	}
+	// get new CTP associations based on destination
+	if len(tp.Destination.IdentityName) > 0 {
+		// Does the identity exist in our mappings?
+		ctp := tm.GetCTPForWorkloadIdentity(tp.Destination.IdentityName)
+		if ctp != nil {
+			tm.mapper.AddLinksForItem(ctp, []resource.ReferenceOrID{res.Id}, false)
```

Comment:
For deletes, this will be called with the pre-delete data. I think it's safer to update the cache after fetching computed traffic permissions in `Reconcile`.

Reply:
I'm not sure I understand the suggestion. Broadly speaking, the mapper should contain the maximal set of possible associations, so that nothing is missed when aggregating and pruning in the reconcile. I see a case where a deleted TP gets mapped and queues up a reconcile, which doesn't end up happening until after a WI is deleted - in which the TP remains unpruned and leaks. In that case, it seems like we should prune all deleted TPs when untracking a WI during reconcile.

But I think for the additive case, we definitely need to put this link in the map before reconcile, no?